File: flink-table-1.7.1/src/main/java/TPCHQuery3Table.java
import org.apache.flink.api.java.DataSet;
import org.apache.flink.api.java.ExecutionEnvironment;
import org.apache.flink.table.api.Table;
import org.apache.flink.table.api.TableEnvironment;
import org.apache.flink.table.api.java.BatchTableEnvironment;
import org.apache.flink.table.descriptors.Csv;
import org.apache.flink.table.descriptors.FileSystem;

import java.sql.Date;

/*
此功能未经过实际运行测试
 */

public class TPCHQuery3Table {
    private static String lineItemPath = null;
    private static String customerPath = null;
    private static String ordersPath = null;
    private static String outputPath = null;
    private static boolean parseParameters(String[] args) {
        if (args.length == 4) {
            lineItemPath = args[0];
            customerPath = args[1];
            ordersPath = args[2];
            outputPath = args[3];
            return true;
        } else {
            return false;
        }
    }
    private static class LineItem {
        public long id;
        public double extdPrice;
        public double discount;
        public String shipDate;
        public LineItem() {}
        public LineItem(long id, double extdPrice, double discount, String shipDate) {
            this.id = id;
            this.extdPrice = extdPrice;
            this.discount = discount;
            this.shipDate = shipDate;
        }
        @Override
        public String toString() {
            return "LineItem{" +
                    "id=" + id +
                    ", extdPrice=" + extdPrice +
                    ", discount=" + discount +
                    ", shipDate='" + shipDate + '\'' +
                    '}';
        }
    }
    private static DataSet<LineItem> getLineItemDataSet(ExecutionEnvironment env) {
        return env.readCsvFile(lineItemPath)
                .fieldDelimiter("|").includeFields("10000110001")
                .pojoType(LineItem.class, "id", "extdPrice", "discount", "shipDate");
    }
    private static class Customer {
        public long id;
        public String mktSegment;
        public Customer() {}
        public Customer(long id, String mktSegment) {
            this.id = id;
            this.mktSegment = mktSegment;
        }
        @Override
        public String toString() {
            return "Customer{" +
                    "id=" + id +
                    ", mktSegment='" + mktSegment + '\'' +
                    '}';
        }
    }
    private static DataSet<Customer> getCustomerDataSet(ExecutionEnvironment env) {
        return env.readCsvFile(customerPath)
                .fieldDelimiter("|").includeFields("1000001")
                .pojoType(Customer.class, "id", "mktSegment");
    }
    private static class Order {
        public long orderId;
        public long custId;
        public String orderDate;
        public long shipPrio;
        public Order() {}
        public Order(long orderId, long custId, String orderDate, long shipPrio) {
            this.orderId = orderId;
            this.custId = custId;
            this.orderDate = orderDate;
            this.shipPrio = shipPrio;
        }
        @Override
        public String toString() {
            return "Order{" +
                    "orderId=" + orderId +
                    ", custId=" + custId +
                    ", orderDate='" + orderDate + '\'' +
                    ", shipPrio=" + shipPrio +
                    '}';
        }
    }
    private static DataSet<Order> getOrdersDataSet(ExecutionEnvironment env) {
        return env.readCsvFile(ordersPath)
                .fieldDelimiter("|").includeFields("11001001")
                .pojoType(Order.class, "orderId", "custId", "orderDate", "shipPrio");
    }
    /*
    USAGE: TPCHQuery3Expression <lineitem-csv path> <customer-csv path> <orders-csv path> <result path>
    */
    public static void main(String[] args) throws Exception {
        if (!parseParameters(args)) {
            return;
        }
        Date date = Date.valueOf("1995-03-12");
        //Environment
        ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
        BatchTableEnvironment tableEnv = TableEnvironment.getTableEnvironment(env);

        //TODO 这里有个问题需要解决：如何在Java中表达Scala的 filter('orderDate.toDate < date)
        Table lineItems = tableEnv.fromDataSet(getLineItemDataSet(env)).filter("TO_DATE(orderDate) < date");
        Table customers = tableEnv.fromDataSet(getCustomerDataSet(env)).filter("mktSegment === AUTOMOBILE");
        Table orders = tableEnv.fromDataSet(getOrdersDataSet(env)).filter("TO_DATE(orderDate) < date");

        Table items = orders.join(customers).where("custId = id").select("orderId, orderDate, shipPrio")
                .join(lineItems).where("orderId = id").select("orderId, extdPrice * (1.0 - discount) as revenue, orderDate, shipPrio");
        Table result = items.groupBy("orderId, orderDate, shipPrio")
                .select("orderId, SUM(revenue) as revenue, orderDate, shipPrio")
                .orderBy("revenue DESC, orderDate ASC");

        //TODO result.writeAsCsv(outputPath, "\n", "|")
        tableEnv.connect(new FileSystem().path(outputPath)).withFormat(new Csv().fieldDelimiter("|"));
        env.execute();
    }
}

File: flink-table-1.7.1/src/main/scala/scala/TPCHQuery3Table.scala
package scala

import org.apache.flink.api.scala.{DataSet, ExecutionEnvironment}
import org.apache.flink.api.scala._
import org.apache.flink.table.api.TableEnvironment
import org.apache.flink.table.api.scala._

object TPCHQuery3Table {
  /*
  USAGE: TPCHQuery3Expression <lineitem-csv path> <customer-csv path> <orders-csv path> <result path>
   */
  def main(args: Array[String]): Unit = {
    if (!parseParameters(args)) {
      return
    }
    val date = "1995-03-12".toDate
    //Environment
    val env = ExecutionEnvironment.getExecutionEnvironment
    val tableEnv = TableEnvironment.getTableEnvironment(env)
    //Source-Table-SQL
    val lineItems = getLineItemDataSet(env)
      .toTable(tableEnv, 'id, 'extdPrice, 'discount, 'shipDate)
      .filter('shipDate.toDate > date)
    val customers = getCustomerDataSet(env)
      .toTable(tableEnv, 'id, 'mktSegment)
      .filter('mktSegment === "AUTOMOBILE")
    val orders = getOrdersDataSet(env)
      .toTable(tableEnv, 'orderId, 'custId, 'orderDate, 'shipPrio)
      .filter('orderDate.toDate < date)
    //SQL
    val items = orders.join(customers).where('custId === 'id).select('orderId, 'orderDate, 'shipPrio)
      .join(lineItems).where('orderId === 'id)
      .select('orderId, 'extdPrice * (1.0f.toExpr - 'discount) as 'revenue, 'orderDate, 'shipPrio)
    val result = items.groupBy('orderId, 'orderDate, 'shipPrio)
      .select('orderId, 'revenue.sum as 'revenue, 'orderDate, 'shipPrio).orderBy('revenue.desc, 'orderDate.desc)
    //DataSet-Sink
    result.writeAsCsv(outputPath, "\n", "|")
    env.execute()
  }
  private var lineItemPath: String = _
  private var customerPath: String = _
  private var ordersPath: String = _
  private var outputPath: String = _
  private def parseParameters(args: Array[String]): Boolean = {
    if (args.length == 4) {
      lineItemPath = args(0)
      customerPath = args(1)
      ordersPath = args(2)
      outputPath = args(3)
      true
    } else {
      false
    }
  }
  case class LineItem(id: Long, extdPrice: Double, discount: Double, shipDate: String)
  private def getLineItemDataSet(env: ExecutionEnvironment): DataSet[LineItem] = {
    env.readCsvFile[LineItem](lineItemPath, fieldDelimiter = "|", includedFields = Array(0, 5, 6, 10))
  }
  case class Customer(id: Long, mktSegment: String)
  private def getCustomerDataSet(env: ExecutionEnvironment): DataSet[Customer] = {
    env.readCsvFile[Customer](customerPath, fieldDelimiter = "|", includedFields = Array(0, 6))
  }
  case class Order(orderId: Long, custId: Long, orderDate: String, shipPrio: Long)
  private def getOrdersDataSet(env: ExecutionEnvironment): DataSet[Order] = {
    env.readCsvFile[Order](ordersPath, fieldDelimiter = "|", includedFields = Array(0, 1, 4, 7))
  }
}
